catalogd/internal/storage/localdir.go

```diff
@@ -2,113 +2,317 @@ package storage
+		if !indexFileStat.Mode().IsRegular() {
+			return false
+		}
+	}
+	return true
+}
+
+func (s *LocalDirV1) catalogDir(catalog string) string {
+	return filepath.Join(s.RootDir, catalog)
+}
+
+func catalogFilePath(catalogDir string) string {
+	return filepath.Join(catalogDir, "catalog.jsonl")
+}
+
+func catalogIndexFilePath(catalogDir string) string {
+	return filepath.Join(catalogDir, "index.json")
+}
+
+type storeMetasFunc func(catalogDir string, metaChan <-chan *declcfg.Meta) error
+
+func storeCatalogData(catalogDir string, metas <-chan *declcfg.Meta) error {
+	f, err := os.Create(catalogFilePath(catalogDir))
+	if err != nil {
+		return err
+	}
+	defer f.Close()
+
+	for m := range metas {
+		if _, err := f.Write(m.Blob); err != nil {
+			return err
+		}
+	}
+	return nil
+}
+
+func storeIndexData(catalogDir string, metas <-chan *declcfg.Meta) error {
+	idx := newIndex(metas)
+
+	f, err := os.Create(catalogIndexFilePath(catalogDir))
+	if err != nil {
+		return err
+	}
+	defer f.Close()
+
+	enc := json.NewEncoder(f)
+	enc.SetEscapeHTML(false)
+	return enc.Encode(idx)
+}
+
+func (s *LocalDirV1) BaseURL(catalog string) string {
 	return s.RootURL.JoinPath(catalog).String()
 }
 
-func (s LocalDirV1) StorageServerHandler() http.Handler {
+func (s *LocalDirV1) StorageServerHandler() http.Handler {
 	mux := http.NewServeMux()
-	fsHandler := http.FileServer(http.FS(&filesOnlyFilesystem{os.DirFS(s.RootDir)}))
-	spHandler := http.StripPrefix(s.RootURL.Path, fsHandler)
-	gzHandler := gzhttp.GzipHandler(spHandler)
 
-	typeHandler := http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
+	mux.HandleFunc(s.RootURL.JoinPath("{catalog}", "api", "v1", "all").Path, s.handleV1All)
+	if s.EnableQueryHandler {
+		mux.HandleFunc(s.RootURL.JoinPath("{catalog}", "api", "v1", "query").Path, s.handleV1Query)
+	}
+	allowedMethodsHandler := func(next http.Handler, allowedMethods ...string) http.Handler {
+		allowedMethodSet := sets.New[string](allowedMethods...)
+		return http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
+			if !allowedMethodSet.Has(r.Method) {
+				http.Error(w, http.StatusText(http.StatusMethodNotAllowed), http.StatusMethodNotAllowed)
+				return
+			}
+			next.ServeHTTP(w, r)
+		})
+	}
+	return allowedMethodsHandler(mux, http.MethodGet, http.MethodHead)
+}
+
+func (s *LocalDirV1) handleV1All(w http.ResponseWriter, r *http.Request) {
+	s.m.RLock()
+	defer s.m.RUnlock()
+
+	catalog := r.PathValue("catalog")
+	catalogFile, catalogStat, err := s.catalogData(catalog)
+	if err != nil {
+		httpError(w, err)
+		return
+	}
+	w.Header().Add("Content-Type", "application/jsonl")
+	http.ServeContent(w, r, "", catalogStat.ModTime(), catalogFile)
+}
+
+func (s *LocalDirV1) handleV1Query(w http.ResponseWriter, r *http.Request) {
+	s.m.RLock()
+	defer s.m.RUnlock()
+
+	catalog := r.PathValue("catalog")
+	catalogFile, catalogStat, err := s.catalogData(catalog)
+	if err != nil {
+		httpError(w, err)
+		return
+	}
+	defer catalogFile.Close()
+
```

Comment:
Think we also need to handle unsupported/unsupportable Range requests here:
```golang
       rtype, haveType := w.Header()["Ranges"]
       if haveType {
               httpError(w, fmt.Errorf("unexpected Range header: %q", rtype))
       }
```

Reply:
To be fair, we do not respond with any kind of `Accept-Range` header, so clients are not supposed to send us `Range` requests, but they might, unless we unconditionally send them a `Accept-Ranges: none` response header.
This can be deferred.  I just wanted to record the note while I was exploring the space.

Reply:
Created https://github.com/operator-framework/operator-controller/issues/1696 to capture follow up